File: agent/src/main/java/com/teamscale/jacoco/agent/testimpact/TestwiseCoverageAgent.java
/*-------------------------------------------------------------------------+
|                                                                          |
| Copyright (c) 2009-2018 CQSE GmbH                                        |
|                                                                          |
+-------------------------------------------------------------------------*/
package com.teamscale.jacoco.agent.testimpact;

import com.teamscale.jacoco.agent.AgentBase;
import com.teamscale.jacoco.agent.GenericExceptionMapper;
import com.teamscale.jacoco.agent.options.AgentOptions;
import com.teamscale.jacoco.agent.util.LoggingUtils;
import com.teamscale.report.testwise.jacoco.JaCoCoTestwiseReportGenerator;
import org.glassfish.jersey.server.ResourceConfig;
import org.glassfish.jersey.server.ServerProperties;
import org.slf4j.Logger;

import java.io.IOException;

/**
 * A wrapper around the JaCoCo Java agent that starts a HTTP server and listens for test events.
 */
public class TestwiseCoverageAgent extends AgentBase {

	/**
	 * The test event strategy handler.
	 */
	protected final TestEventHandlerStrategyBase testEventHandler;

	/** Creates a {@link TestwiseCoverageAgent} based on the given options. */
	public static TestwiseCoverageAgent create(AgentOptions agentOptions) throws IOException {
		Logger logger = LoggingUtils.getLogger(JaCoCoTestwiseReportGenerator.class);
		JaCoCoTestwiseReportGenerator reportGenerator = new JaCoCoTestwiseReportGenerator(
				agentOptions.getClassDirectoriesOrZips(), agentOptions.getLocationIncludeFilter(),
				agentOptions.getDuplicateClassFileBehavior(), LoggingUtils.wrap(logger));
		return new TestwiseCoverageAgent(agentOptions,
				new TestExecutionWriter(agentOptions.createNewFileInOutputDirectory("test-execution", "json")),
				reportGenerator);
	}


	public TestwiseCoverageAgent(AgentOptions options, TestExecutionWriter testExecutionWriter,
			JaCoCoTestwiseReportGenerator reportGenerator) throws IllegalStateException {
		super(options);
		switch (options.getTestwiseCoverageMode()) {
			case TEAMSCALE_UPLOAD:
				testEventHandler = new CoverageToTeamscaleStrategy(controller, options, reportGenerator);
				break;
			case DISK:
				testEventHandler = new CoverageToDiskStrategy(controller, options, reportGenerator);
				break;
			case HTTP:
				testEventHandler = new CoverageViaHttpStrategy(controller, options, reportGenerator);
				break;
			default:
				testEventHandler = new CoverageToExecFileStrategy(controller, options, testExecutionWriter);
				break;
		}
	}

	@Override
	protected ResourceConfig initResourceConfig() {
		ResourceConfig resourceConfig = new ResourceConfig();
		resourceConfig.property(ServerProperties.WADL_FEATURE_DISABLE, Boolean.TRUE.toString());
		TestwiseCoverageResource.setAgent(this);
		return resourceConfig.register(TestwiseCoverageResource.class).register(GenericExceptionMapper.class);
	}

	@Override
	public void dumpReport() {
		// Dumping via the API is not supported in testwise mode. Ending the test run dumps automatically
	}
}

File: agent/src/main/java/com/teamscale/jacoco/agent/testimpact/TestwiseCoverageAgent.kt
/*-------------------------------------------------------------------------+
|                                                                          |
| Copyright (c) 2009-2018 CQSE GmbH                                        |
|                                                                          |
+-------------------------------------------------------------------------*/
package com.teamscale.jacoco.agent.testimpact

import com.teamscale.jacoco.agent.AgentBase
import com.teamscale.jacoco.agent.AgentOptions
import com.teamscale.jacoco.agent.JacocoRuntimeController.DumpException
import spark.Request
import spark.Response

import spark.Spark.get
import spark.Spark.port
import spark.Spark.post
import spark.Spark.stop

/**
 * A wrapper around the JaCoCo Java agent that starts a HTTP server and listens for test events.
 */
class TestwiseCoverageAgent
/** Constructor.  */
@Throws(IllegalStateException::class)
constructor(
    /** The agent options.  */
    private val options: AgentOptions
) : AgentBase(options) {

    init {
        initServer()
    }

    /**
     * Starts the http server, which waits for information about started and finished tests.
     */
    private fun initServer() {
        logger.info("Listening for test events on port {}.", options.httpServerPort)
        port(options.httpServerPort!!)

        get("/test") { _, _ -> controller.sessionId }

        post("/test/start/$TEST_ID_PARAMETER") { request, _ ->
            handleTestStart(request)
            "success"
        }

        post("/test/end/$TEST_ID_PARAMETER") { request, _ ->
            handleTestEnd(request)
            "success"
        }
    }

    /** Handles the start of a new test case by setting the session ID.  */
    private fun handleTestStart(request: Request) {
        logger.debug("Start test " + request.params(TEST_ID_PARAMETER))

        // Dump and reset coverage so that we only record coverage that belongs to this particular test case.
        controller.reset()
        val testId = request.params(TestwiseCoverageAgent.TEST_ID_PARAMETER)
        controller.sessionId = testId
    }

    /** Handles the end of a test case by resetting the session ID.  */
    @Throws(DumpException::class)
    private fun handleTestEnd(request: Request) {
        logger.debug("End test " + request.params(TEST_ID_PARAMETER))
        controller.dump()
    }

    override fun prepareShutdown() {
        stop()
    }

    companion object {

        /** Path parameter placeholder used in the http requests.  */
        private val TEST_ID_PARAMETER = ":testId"
    }
}
